counterpartylib/lib/messages/versions/mpma.py

```diff
@@ -0,0 +1,434 @@
+#! /usr/bin/python3
+
+import struct
+import json
+import logging
+import binascii
+import math
+import itertools
+import operator
+from bitcoin import base58
+from bitcoin.core import key
+from functools import reduce
+
+from bitstring import BitArray, BitStream, ConstBitStream, ReadError
+
+logger = logging.getLogger(__name__)
+
+from counterpartylib.lib import (config, util, exceptions, util, message_type, address)
+
+ID = 3 # 0x03 is this specific message type
+
+## address manipulation functions
+
+def _addrToBytes(addr):
+    return base58.decode(addr)[:-4]
+
+def _addressFrom21Bytes(b):
+    sha256_1 = key.hashlib.new('sha256')
+    sha256_1.update(b)
+
+    sha256_2 = key.hashlib.new('sha256')
+    sha256_2.update(sha256_1.digest())
+
+    chk = sha256_2.digest()[:4]
+
+    return base58.encode(b''.join([b, chk]))
+
+## encoding functions
+
+def _encode_constructBaseLUT(snds):
+    # t is a tuple of the form (asset, addr, amnt, memo, is_hex)
+    return sorted(list(set([t[1] for t in snds]))) # Sorted to make list determinist
+
+def _encode_constructBaseAssets(snds):
+    # t is a tuple of the form (asset, addr, amnt [, memo, is_hex])
+    return sorted(list(set([t[0] for t in snds]))) # Sorted to make list determinist
+
+def _encode_constructLUT(snds):
+    baseLUT = _encode_constructBaseLUT(snds)
+    lutNbits = math.ceil(math.log2(len(baseLUT)))
+    return {
+        "nbits": lutNbits,
+        "addrs": baseLUT
+    }
+
+def _encode_compressLUT(lut):
+    return b''.join([struct.pack('>H', len(lut['addrs']))] +
+        [
+            _addrToBytes(addr)
+            for addr in lut['addrs']
+        ])
+
+def _encode_memo(memo=None, is_hex=False):
+    if memo is not None:
+        barr = BitArray('0b1')
+
+        if is_hex:
+            barr.append('0b1')
+            if type(memo) is str: # append the string as hex-string
+                barr.append('uint:6=%i' % (len(memo) >> 1))
+                memo = '0x%s' % memo
+            else:
+                barr.append('uint:6=%i' % len(memo))
+
+            barr.append(memo)
+        else:
+            barr.append('0b0')
+            barr.append('uint:6=%i' % len(memo))
+            barr.append(BitArray(memo.encode('utf-8')))
+
+        return barr
+    else:
+        return BitArray('0b0')
+
+def _safe_tuple_index(t, i):
+    if len(t) <= i:
+        return None
+    else:
+        return t[i]
+
+def _encode_constructSendList(send_asset, lut, sends):
+    # t is a tuple of the form (asset, addr, amnt, memo, is_hex)
+    return [
+        (lut['addrs'].index(t[1]), t[2], _safe_tuple_index(t, 3), _safe_tuple_index(t, 4))
+        for t in sends
+        if t[0] == send_asset
+    ]
```

Comment:
So, ```_safe_tuple_index``` exists to check whether there is a memo or not? Because there are so many steps in creating and validating this kind of send, I think this code would benefit from a lot more comments.